plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsStepContextTest.java

```diff
@@ -0,0 +1,203 @@
+    public void failingStepListenerNotLeakClosures() throws Exception {
+        // Even before the fix there's only one warning logged.  Asserting zero records is probably over-stepping,
+        // but asserting just one record with our target message risks a false negative (some other unrelated message
+        // being first, and our being later).
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("node {}\n", true));
+
+        WorkflowRun build = r.buildAndAssertStatus(Result.FAILURE, job);
+        r.assertLogContains("oops", build);
+        assertThat(ClosureCounter.get().closureCount, equalTo(0));
+        assertThat(logger.getMessages(), not(hasItem(containsString("Stale closure"))));
+    }
+
+    @TestExtension("failingStepListenerNotLeakClosures")
+    public static class FailingStepListener implements StepListener {
+
+        @Override
+        public void notifyOfNewStep(@NonNull Step s, @NonNull StepContext context) {
+            context.onFailure(new AbortException("oops"));
+        }
+    }
+
+    @Issue("JENKINS-75067")
+    @Test
+    public void executionStartExceptionNotLeakClosures() throws Exception {
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("badBlock {}\n", true));
+
+        WorkflowRun build = r.buildAndAssertStatus(Result.FAILURE, job);
+        r.assertLogContains("oops", build);
+        assertThat(ClosureCounter.get().closureCount, equalTo(0));
+        assertThat(logger.getMessages(), not(hasItem(containsString("Stale closure"))));
+    }
+
+    @Issue("JENKINS-75067")
+    @Test
+    public void executionWithBodyRunningSyncNotLeakClosures() throws Exception {
+        logger.record(CpsThreadGroup.class, Level.WARNING).capture(10);
+        WorkflowJob job = r.createProject(WorkflowJob.class, "p");
+        job.setDefinition(new CpsFlowDefinition("echo passthrough {}\n", true));
+
+        WorkflowRun build = r.buildAndAssertSuccess(job);
+        r.assertLogContains("hooray", build);
+        assertThat(ClosureCounter.get().closureCount, equalTo(0));
+        assertThat(logger.getMessages(), not(hasItem(containsString("Stale closure"))));
+    }
+
+    public static class BadBlockStep extends Step {
+
+        @DataBoundConstructor
+        public BadBlockStep() {}
+
+        @Override
+        public StepExecution start(StepContext context) throws Exception {
+            return new Execution(context);
+        }
+
+        public static class Execution extends StepExecution {
+
+            public Execution(StepContext context) {
+                super(context);
+            }
+
+            @Override
+            public boolean start() throws Exception {
+                throw new RuntimeException("oops");
+            }
```

Comment:
Just FYI there are some helpers that can be used to simplify implementations of `StepExecution` in tests:

```suggestion
        return StepExecutions.synchronous(context, ctx -> {
            throw new AbortException("oops");
        });
```

Reply:
`throw` is a statement, so the braces are needed here

Reply:
Thanks for the reminders!  I was aware of them when I originally set out to create my own plugin, but they were all to simplistic for my actual needs, so I forgot about them (and they didn't jump out at me in the existing tests I looked at).  But they are indeed perfect for these simple tests.